server/update-workflow-interface/update-workflow-interface.md

```diff
@@ -0,0 +1,217 @@
+completed, subject to history storage & archival configuration.
+
+## Use Cases
+
+### Short Update
+
+It may be that it is known a priori that an update is expected to execute in an
+amount of time that is reasonable for an RPC request/response interaction. In
+such cases the issuer may want to block the issuing thread of execution for the
+full duration of the update.
+
+### Uncertain Duration Update
+
+It may be the case that there is no expectation on how long an update may take
+or the duration may be inherently variable. In such cases the client may want to
+block for _up to_ a client-chosen duration after which time the issuing client
+disconnects from the RPC but the update processing continues on. The system must
+support the client re-attaching to the Update at some point in the future in
+order to observe the UpdateResponse. Note that this is different than putting a
+timeout on the RPC itself.
+
+### Known Long Update
+
+Similar to the uncertain duration, it may be know a priori that the update will
+take more time than the issuer is willing to block the thread of execution
+invoking the update. In such cases the update issuer would prefer to not wait at
+all for the UpdateResponse, preferring instead to poll for it later.
+
+### Invalid Update
+
+UpdateRequests can be invalid given the state of the receiving workflow. In such
+cases the UpdateResponse will contain an error indication and payload. Invalid
+updates _must not_ be recorded in workflow history to prevent a malicious issuer
+from causing a denial of service on the workflow by issuing repeated invalid
+UpdateRequests.
+
+### "Failing" Update
+
+UpdateRequests can be semantically acceptable but still fail due to external
+dependencies. In such cases the UpdateResponse will contain an error indication
+and payload.
+
+### Unexpected Disconnect
+
+While the issuer of an update may wish to block on an UpdateResponse, network
+conditions or software failures may occur which forcibly disconnects the update
+issuer from the server at the time that the Update completes _and_ before the
+issuer becomes aware of the Update ID. The issuer must be able to re-issue the
+same UpdateRequest to access the server-generated Update ID.
+
+## gRPC API
+
+The gRPC API comprises two functions, `UpdateWorkflow` and
+`AwaitUpdateWorkflowResponse` to be added to the `WorkflowService` service
+definition. The former is used to invoke an update and follows the standard
+patterns for identifying a workflow execution and carrying arbitrary metadata
+(Header) and payloads (Payloads). The request type for this function includes a
+`wait_timeout` property indicating how long the client is willing to wait for a
+synchronous response once the request has been accepted by the system. This
+value can be set to zero indicating no such waiting should occur.  The
+`request_id` property uniquely identifies the logical request (_not_ the
+issuance of the request) and is intended for use as an idempotency token.
```

Comment:
Is this `request_id` alone what is used for deduplication? Or is it combined with other values?

If `request_id` is the idempotency token for this purpose, can it not be used as the update ID? It seems much simpler to have a user-provided update ID that is used for deduplication and can be used everywhere to reference the update.

Reply:
I've sort of mentioned this obliquely in some other responses but the idea is that if there is an update ID then the update will (eventually) happen and there will (eventually) be an UpdateResponse. For that to be the case, it has to be generated on the server side _after_ the workflow has validated the UpdateRequest.

Reply:
P.S. I'm noting that this arrangement (request_id vs update_id and the implications) wasn't clear in the proposal and I'll improve that.

Reply:
You're good here, I got confused in thinking this was a user-facing construct for them to be idempotent with their updates. But really this is internal idempotency between client and server like out other request IDs.